js/id/ui/preset/wikipedia.js

```diff
@@ -105,9 +107,23 @@ iD.ui.preset.wikipedia = function(field, context) {
             title.value(value);
         }
 
-        var t = {};
-        t[field.key] = value ? language()[2] + ':' + value : undefined;
-        event.change(t);
+        syncTags[field.key] = value ? language()[2] + ':' + value : undefined;
+        event.change(syncTags);
+        
+        if (value && language()[2]) {
+            wikidata.itemsByTitle(language()[2], value, function (title, entities) {
+                var qids = entities && Object.keys(entities),
+                    qid = qids && _.find(qids, function (id) {
+                    return id.match(/^Q\d+$/);
+                }),
+                    selectedIDs = context.selectedIDs && context.selectedIDs();
+                if (selectedIDs && selectedIDs.length === 1 && selectedIDs[0] === entity.id) {
+                    event.change({wikidata: qid});
+                } else {
+                    context.entity(entity.id).tags.wikidata = qid;
```

Comment:
I'm not going to claim that my approach is correct :)  but I created an action and then had the context 'perform' it. See https://github.com/strava/iD/blob/master/js/id/operations/slide.js#L103  

This looks like it edits the context/graph data directly, which, as has been said above, messes with the immutability of the context. (This feature makes it easy to do undo/redo as you can just swap contexts).

I was assuming that calling the perform on the context took care of serializing the edits and updating the history (it's seemed to work in practice). The worry is the context/graph may have changed in between the time you clicked the button and when the request came back. 

I hope this is helpful.